### DISCUSSION TITLE
does proof verification of a group asset need more calculation power?

### DISCUSSION THREAD

**Discussion by pynixwang** [lightninglabs/taproot-assets#826] [Q&A]
my teams are very very good at creating `large group assets` projects(for example 600k NFTs), I want to know:

1, is large group assets taking more time to verify
2, is proof verification taking more time than networking or database
3, what is rate limit of pushing proof to official universe.
4, how to deal with large group assets

**Comment by guggero** [comment]
Hi. Thank you for your question.

1. Because a non-fungible asset (called "collectible" in this project) will always have the amount of `1`, you are correct that you will have to mint them into an asset group if they should be assigned to the same collection of NFTs. Because of that, each asset will have its own asset ID and proof, so it will take quite a while to **mint** them. We have done up to 10k assets in our tests and that can take up to an hour in some cases. The most time there is spent uploading the proofs to the universe. BUT, that's a time investment you only have to do once.
2. I'm not sure I understand the question correctly. I interpret it as "Which part of processing an asset is the most time intensive: the proof verification, the proof download or the database interaction". The answer would be: it depends.
   1.  When **minting**, the database insertion of the proofs both to the local universe (on the node that does the mint) and the remote universe (whichever node you choose to serve as a central repository for the proofs, can be the same node as the minting one) is what takes most of the time. Proofs are quite small (up to 1 MB of optional metadata plus a few kilobytes per transaction), so networking isn't really that big of an issue (unless you create 600k NFTs with 1 MB of image data each, then you'll of course need to upload that to the universe). The actual proof verification is quite "cheap" in terms of CPU cycles. I would say in the order of a few milliseconds per proof.
   2. When **receiving** an asset, all three parts are fairly small: Download of around 1MB (if there is metadata), a few milliseconds of proof verification and a few milliseconds of inserting the proof into the local database.
3. The rate limits can be configured by each universe. The default values are here: https://github.com/lightninglabs/taproot-assets/blob/9cd7d53c19611d41f3ebf8522d1467f54d3b289d/tapcfg/config.go#L121
4. Check our example load test script that creates up to 10k assets in a single group: https://github.com/lightninglabs/taproot-assets/blob/8d991370b15fc6b0534f752c7fcda1abcfcfeae6/itest/mint_batch_stress_test.go#L47
